---
title: "Analysis of TCGA PAAD tumor sample RNA-seq data for HIF/EMT enrichment (PDAC tummors only)"
author: Paul J. Myers
output:
  html_document:
    fig_width: 9
    fig_height: 6
    theme: spacelab
    toc: yes
  pdf_document: 
    toc: yes
    fig_width: 9
    fig_height: 6
Date of creation: "12/08/2020"
R version: "4.1.2"
---

# Housekeeping Code
We start by loading some of the desired libraries for this script. Additional packages will be loaded later for specific analyses to make their use clear.

```{r Load packages, message=F, warning=F}
### Load packages:
library(tidyverse)
library(magrittr)
library(BiocManager)
library(paletteer)
library(cowplot)
library(ggstatsplot)
library(ggbeeswarm)
library(tidyHeatmap)
library(ComplexHeatmap)
library(parallel)
```

```{r Clear workspace}
### Clear R workspace:
rm(list=ls())
```

```{r Check working directory}
### Check that current working directory is correct and the desired one:
cwd <- getwd()
cwd
```

```{r Define ni operator}
## Define 'not in' (ni) operator
`%ni%` <- Negate(`%in%`)
```

This function can be used to generate a ggplot with a blank grid and black box surrounding the plot area. The only input to the function specifies whether y tick marks and labels are desired or not (depending on the type of plot you are generating).
```{r Custom ggplot theme for plot background}
theme_cust <- function(ticks=T, axis="y", box=T, lnsz=1,
                       xtickangle=0, hjust=NULL, vjust=NULL, fill=NA,
                       legend.position="right"){
  theme_out <- theme(legend.position=legend.position)
  if (ticks == FALSE & axis == "y" ){
    theme_out <- theme_out + theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = fill),
      )
  } else if (ticks == FALSE & axis != "y"){
    theme_out <- theme_out + theme(
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = fill),
      )
  } else {
     theme_out <- theme_out + theme(
       axis.line.y = element_blank(),
       axis.line.x = element_blank(),
       panel.background = element_rect(size = lnsz, color = "black", fill = fill),
       axis.text.x = element_text(angle=xtickangle, hjust = hjust, vjust=vjust),
       ) 
  }
  
  if (xtickangle==90){
    theme_out <- theme_out + theme(axis.text.x = element_text(angle=xtickangle, hjust = 1, vjust=0.4))
  }
  
  if (box != TRUE) {
    theme_out <- theme_out + theme(panel.background = element_blank(),
                                   axis.line.y = element_line(size = lnsz, color = "black"),
                                   axis.line.x = element_line(size = lnsz, color = "black")
                                   )
  } else {
    theme_out <- theme_out
  }
  return(theme_out)
}

# Julia's Plots.jl default color order:
julia_cols <- c("#009AFA","#E26E47","#3FA54E","#c270D2","#AD8F18","#01ABAE","#ED5F92","#C68324","#01A98C","#8F961E","#01A9CD","#9B7EE8","#618CF7","#F16072","#DC65B7","#6D9E33")
```

# Load data
The TCGA Pancreatic Cancer  data were downloaded from the [UCSC Xena Portal](https://xenabrowser.net/datapages/?cohort=GDC%20TCGA%20Pancreatic%20Cancer%20&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443). Per the Xena website:
"The gene expression profile was measured experimentally using the Illumina HiSeq 2000 RNA Sequencing platform by the University of North Carolina TCGA genome characterization center. Level 3 data was downloaded from TCGA data coordination center. This dataset shows the gene-level transcription estimates, as in log2(x+1) transformed RSEM normalized count. Genes are mapped onto the human genome coordinates using UCSC Xena HUGO probeMap."

Clinical drug treatment data were downloaded from the [TCGA online portal](https://portal.gdc.cancer.gov/). Phenotype and curated survival data were downloaded from the [UCSC Xena Portal](https://xenabrowser.net/datapages/?cohort=TCGA%20Pancreatic%20Cancer%20&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443). For the survival data, the following abbreviations are used:
* OS = overall survival
* PFI = progression-free interval
* DSS = disease-specific survival
* DFI = disease-free interval

```{r Clinical + phenotype + mutation data, message=F, warning=F}
## =============== Clinical/phenotype data =============== ## 
fn.clin <- "TCGA PAAD data/PAAD_clinicalMatrix.txt"
clin_data <- read_tsv(fn.clin) %>% 
  filter(str_detect(histological_type,"Ductal"))
colnames(clin_data) <- colnames(clin_data) %>% make.names()


## =============== Drug data =============== ## 
fn.drug <- "TCGA PAAD data/clinical_drug_tcga_paad.txt"  # file name
drug_data <- read_tsv(fn.drug)[-1,] %>%  # load data
  mutate(drug_name = tolower(pharmaceutical_therapy_drug_name)) %>% 
  filter(bcr_patient_barcode %in% clin_data$`_PATIENT`)
  

## =============== Curated survival data =============== ## 
fn.surv <- "TCGA PAAD data/PAAD_survival.txt"
surv_data <- read_tsv(fn.surv)
colnames(surv_data) <- colnames(surv_data) %>% make.names()
surv_data <- surv_data %>% filter(X_PATIENT %in% clin_data$X_PATIENT)


## =============== Mutation data =============== ## 
fn.mut1 <- "TCGA PAAD data/PAAD_mc3.txt"
mut_data <- read_tsv(fn.mut1) %>% 
  filter(sample %in% clin_data$sampleID)
colnames(mut_data) <- colnames(mut_data) %>% make.names()

# Get KRAS mutation status and add to sample metadata:
kras_mut <- mut_data %>%
  filter(gene=="KRAS" & str_detect(Amino_Acid_Change,"G12") & effect=="Missense_Mutation") %>%
  dplyr::select(sample, Amino_Acid_Change) %>% 
  group_by(sample) %>% 
  summarise(KRAS_mut=paste(Amino_Acid_Change, collapse=", ")) %>% 
  dplyr::rename(
    sample.id=sample,
    )


## =============== Gene-level mutation calls =============== ##
fn.mut2 <- "TCGA PAAD data/PAAD_mc3_gene_level.txt"
gmut_data <- read_tsv(fn.mut2) %>% 
  dplyr::rename(gene=sample)
colnames(gmut_data) <- colnames(gmut_data) %>% make.names()


```

```{r RNA-seq data, message=F}
## Load data:
fn <- "TCGA PAAD data/HiSeqV2.txt"
raw.data <- read_tsv(fn) %>% dplyr::rename(gene=sample) # load the data -- given as log2(x+1) RSEM normalized counts
data <- data.frame(raw.data) # make a copy for use downstream
colnames(data)[-1] <- paste("SPL",1:length(colnames(data)[-1]), sep="")
rownames(data) <- data$gene

spls <- data.frame(sample.id = colnames(raw.data)[-1],# %>% sub("-01$|-11$|-06$","",.), 
                   spl = colnames(data)[-1]) %>% 
  mutate(patient.id = gsub("(.*)-.*", "\\1", sample.id)) %>% 
  left_join(kras_mut, by="sample.id") %>% 
  mutate(
    KRAS_mut = replace(KRAS_mut, is.na(KRAS_mut), "WT"),
    KRAS_mut = gsub("p.","",KRAS_mut)
    ) %>% 
  filter(sample.id %in% clin_data$sampleID)
rownames(spls) <- spls$spl

## Get RSEM, TPM, and log2(TPM+1) count data:
rsem <- data[, c("gene",spls$spl)]; rsem[,-1] <- 2^rsem[,-1]-1 # convert to RSEM normalized counts, keeping only PDAC samples
rownames(rsem) <- rsem$gene

tpm <- rsem[,-1] %>% # Get just numeric entries (ignore Gene_ID)
  data.matrix() %>% # Convert to numeric matrix
  scale(center=F, scale=colSums(.)) %>% # Divide columns by sum of total transcripts (RSEM normalized counts)
  multiply_by(1.0e6) %>%   # Multiply by one million (now have TPM)
  data.frame(row.names =rsem$gene, gene = rsem$gene, .)

logtpm <- log2(tpm[,-1]+1) %>% 
  data.frame(row.names = rsem$gene, gene = rsem$gene, .) # convert back to data frame


```
# Immunedeconv
Here we attempt to use deconvolution methods from the `immunedeconv` package to estimate fractions of tumor and immune cells in the bulk TCGA data. `immunedeconv` collects several different deconvolution methods into a single framework that makes using them much easier and more straightforward. We will look at using the methods that provide absolute estimates of immune cell and uncharacterized cell (tumor cell) abundances in the form of cell type fractions.

For detailed installation and usage instructions for `immunedeconv`, see the [package website](https://icbi-lab.github.io/immunedeconv/index.html).

```{r Installing immunedeconv}
# install.packages("remotes") # install if needed
# remotes::install_github("icbi-lab/immunedeconv") # install if needed
```

`immunedeconv` recommends that gene expression be provided in *non-log-transformed* transcripts per million (TPM) units unless using the xCell and MCP-counter algorithms, which are not as reliant on following this convention.

```{r Run immunedeconv}
library(immunedeconv)

############ Deconvolution with EPIC algorithm ############
deconv1 <- tpm %>% dplyr::select(-gene) %>%
  deconvolute(method = "epic", tumor = T)

```

Now viewing the results from immune deconvolution.
```{r View immunedeconv results, message = F, warning=F}
## Results w/ EPIC:
spl_deconv_ord <- deconv1 %>%
  pivot_longer(-cell_type) %>%
  filter(cell_type=="uncharacterized cell") %>%
  arrange(value) %>% mutate(name=factor(name,levels=name)) %>%
  pull(name) # get order of samples based on decreasing tumor cell content

deconv1 %>%
  pivot_longer(-cell_type, names_to = "sample", values_to = "fraction") %>%
  mutate(sample = factor(sample, levels=spl_deconv_ord)) %>%
  ggplot(aes(x=sample, y=fraction, fill=cell_type)) +
    geom_bar(stat="identity") +
    coord_flip() +
    scale_fill_brewer(palette = "Paired") +
    scale_y_continuous(expand=expansion(mult=c(0, 0))) +
    theme_minimal_vgrid(13) + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
    labs(title="TCGA PAAD RNA-seq\nImmune deconvolution w/ EPIC algorithm",
         x = "Sample", y="Fraction")
ggsave("Images and plots/immune-deconv_results_EPIC_PDAC-only.png", height=7)

deconv1 %>%
  filter(cell_type=="uncharacterized cell") %>%
  pivot_longer(-cell_type, names_to = "sample", values_to = "tumor_cell_fraction") %>%
  gghistostats(tumor_cell_fraction, binwidth=0.01, title="TCGA PAAD: Estimated tumor cell content")
ggsave("Images and plots/immune-deconv_tumor-cell-frac_histogram_EPIC_PDAC-only.png")

```

Finally, we store the deconvolution results with the sample information for use in later calculations. For now, we'll just store the results from the EPIC algorithm.
```{r Store immunedeconv results}
## Add cell fractions to sample information:
spls <- deconv1 %>% pivot_longer(-cell_type) %>%
  pivot_wider(name, values_from = value, names_from = cell_type) %>%
  inner_join(x=spls, y=., by=c("spl"="name"), suffix = c("",".x")) %>%
  mutate(immune_frac = 1-`uncharacterized cell`) %>%
  select_at(vars(-ends_with(c(".x",".y"))))
colnames(spls) %<>% make.names()

hipurity <- spls %>% filter(uncharacterized.cell>0) %>% pull(spl)

```


# Gene signatures of interest
Next we introduce the "HIF signature" -- a set of 44 HIF target genes that were described/used by [Li et al. (2014)](https://doi.org/10.1038/nature13557).
```{r HIF signature}
hif.sig <- c("IGFBP3", "EDN2", "PFKFB4", "FLT1", "TFR2", "BNIP3L", "TGFA","BNIP3","PGK1","EGLN1","LDHA","EGLN3","CP","TGFB3","PFKFB3",
             "HK1","TFRC","EDN1","CDKN1A","CA9","ADM1","HMOX1","SERPINE1","LOX","NDRG1","CA12","PDK1","VEGFA","ERO1L","RORA","P4HA1","MXI1",
             "SLC2A1","STC2","MIF","DDIT4","ENO1","CXCR4","PLOD1","P4HA2","GAPDH","PGAM1","TMEM45A","PIM1")

## log2(TPM+1):
logtpm.hif <- match(hif.sig, logtpm$gene) %>% na.omit() %>% logtpm[.,] %>% data.frame(row.names=.$gene)
logtpm.hifscore <- logtpm.hif %>% dplyr::select(-gene) %>% colMeans()

```

Next we load the *Collisson PDAssigner signature*. This signature was originally defined by [Collisson et al.](https://doi.org/10.1038/nm.2344) and was used by [Porter et al.](https://doi.org/10.1073/pnas.1914915116) to classify the TCGA PAAD samples.
```{r Collisson PDAssigner signature, message=F}
## Load subtypes gene list:
fn.sub <- "pdassigner_genes.csv"
pda_genes <- read_csv(fn.sub)
subtype_anno <- pda_genes[,-1] %>% data.frame()
rownames(subtype_anno) <- pda_genes$gene

pda_class <- pda_genes %>% filter(type=="classical") %>% dplyr::pull(gene)
pda_exo <- pda_genes %>% filter(type=="exocrine_like") %>% dplyr::pull(gene)
pda_qm <- pda_genes %>% filter(type=="QM") %>% dplyr::pull(gene)
pda_sigs <- list(COLLISSON_CLASSICAL = pda_class,
                 COLLISSON_EXOCRINE = pda_exo,
                 COLLISSON_QUASIMESENCHYMAL = pda_qm)

## Pull out the subtype/signature genes:
sub <- logtpm %>% filter(gene %in% pda_genes$gene)
rownames(sub) <- sub$gene
```

```{r Moffitt signatures, message=F}
## Load the signatures:
fn.mof <- "Moffitt_PDAC_signature.txt" # PDAC subtype signature
fn.mof_strom <- "Moffitt_stromal_signature.txt" # stromal signature
mof.sig <- read_tsv(fn.mof)
mof_strom.sig <- read_tsv(fn.mof_strom)

## Get genes from subsignatures for GSVA calculations:
mof_class <- mof.sig %>% filter(type=="Classical") %>% pull(gene)
mof_bl <- mof.sig %>% filter(type=="Basal-like") %>% pull(gene)
mof_strom_norm <- mof_strom.sig %>% filter(type=="Normal_stroma") %>% pull(gene)
mof_strom_act <- mof_strom.sig %>% filter(type=="Activated_stroma") %>% pull(gene)

mof_sigs <- list(MOFFITT_CLASSICAL = mof_class,
                 MOFFITT_BASAL_LIKE = mof_bl,
                 MOFFITT_NORMAL_STROMA = mof_strom_norm,
                 MOFFITT_ACTIVATED_STROMA = mof_strom_act) # list for GSVA calculations

## Expression of Moffitt signatures:
mof <- logtpm %>% filter(gene %in% mof.sig$gene)
mof_strom <- logtpm %>% filter(gene %in% mof_strom.sig$gene)
```

```{r Bailey signature, message=F}
fn.bail <- "Bailey_PDAC_signature.txt"
bail.sig <- read_tsv(fn.bail)

bail_sq <- bail.sig %>% filter(type=="Squamous") %>% pull(gene)
bail_prog <- bail.sig %>% filter(type=="Progenitor") %>% pull(gene)
bail_adex <- bail.sig %>% filter(type=="ADEX") %>% pull(gene)
bail_imm <- bail.sig %>% filter(type=="Immunogenic") %>% pull(gene)

bail_sigs <- list(BAILEY_SQUAMOUS = bail_sq,
                  BAILEY_PROGENITOR = bail_prog,
                  BAILEY_ADEX = bail_adex,
                  BAILEY_IMMUNOGENIC = bail_imm) # list for GSVA calculations

## Gene expression from Bailey signature:
bail <- logtpm %>% filter(gene %in% bail.sig$gene)
```


## Pan-cancer EMT signature
Here we introduce the pan-cancer EMT (pcEMT) signature from [Mak et al. (2016)](https://doi.org/10.1158/1078-0432.ccr-15-0876). This signature/gene set contains genes associated with the epithelial and mesenchymal cell states, as derived from a pan-cancer analysis of TCGA RNA-seq data in which four marker genes--CDH1 (E), CDH2 (M), VIM (M), and FN1 (M)-- were used as "seed" genes against which the highest correlated genes were identified across tumors and extracted as part of the signature. *In total, the signature consists of 77 genes, 52 of which are associated with the mesenchymal cell state and 25 of which the epithelial cell state.*

```{r Get pcEMT signature, warning=F, message=F}
## Load pan-cancer EMT signature:
pcemt.fn <- "Pan-cancer-EMT-signature_Mak-et-al-2016.txt"
pcemt.sig <- read_tsv(pcemt.fn)
pcemt.sig.m <- read_tsv(pcemt.fn) %>% subset(type %in% "Mesenchymal")

## Get pcEMT signature data:
pcemt <- pcemt.sig$gene %>% match(logtpm$gene) %>% logtpm[.,]
rownames(pcemt) <- pcemt$gene
pcemt.tpm <- pcemt.sig$gene %>% match(tpm$gene) %>% tpm[.,]
rownames(pcemt.tpm) <- pcemt.tpm$gene

## Modify HIF signature to have the same fields as the pcEMT signature:
hif.sig_df <- data.frame(gene = hif.sig, type = "HIF_sig") 
```


## MSigDB gene sets
Here we load the necessary packages and gene sets from/related to the Molecular Signatures Database (MSigDB) for analyzing pathway and gene set enrichment in downstream analyses.
```{r Load packages and MSigDB collections, message=FALSE, warning=FALSE}
## Load packages:
# library(clusterProfiler)
# options(connectionObserver = NULL)
library(org.Hs.eg.db)
library(msigdbr)


### Define MSigDB gene set collection(s) to use --> retrieve with 'msigdbr' package:
species = "Homo sapiens"
 
## Retrieve Hallmark and canonical pathways collections in the database (C1-H and C2-CP):
hall = msigdbr(species = species, category = "H")
cp = msigdbr(species = species, category = "C2", subcategory = "CP")
cp.b = msigdbr(species = species, category = "C2", subcategory = "CP:BIOCARTA")
cp.r = msigdbr(species = species, category = "C2", subcategory = "CP:REACTOME")
cp.p = msigdbr(species = species, category = "C2", subcategory = "CP:PID")
cp.k = msigdbr(species = species, category = "C2", subcategory = "CP:KEGG")
cp.w = msigdbr(species = species, category = "C2", subcategory = "CP:WIKIPATHWAYS")
gene_sets1 <- rbind(hall, cp, cp.b, cp.r, cp.p, cp.k, cp.w) %>% split(x = .$gene_symbol, f = .$gs_name)

## Go collections (C5):
go.bp <- msigdbr(species = species, category = "C5", subcategory = "GO:BP")
go.cc <- msigdbr(species = species, category = "C5", subcategory = "GO:CC")
go.mf <- msigdbr(species = species, category = "C5", subcategory = "GO:MF")
gene_sets2 <- rbind(go.bp, go.cc, go.mf) %>% split(x = .$gene_symbol, f = .$gs_name)

## Transcription factor target collections (C3-TFT):
tft <- msigdbr(species = species, category = "C3", subcategory = "TFT:GTRD")
tft_leg <- msigdbr(species = species, category = "C3", subcategory = "TFT:TFT_Legacy")
gene_sets3 <- rbind(tft, tft_leg) %>% split(x = .$gene_symbol, f = .$gs_name)

## Pan-cancer EMT sub-signatures and HIF signature:
pcemt_e.sig <- pcemt.sig %>% subset(type=="Epithelial") %>% dplyr::pull(gene)
pcemt_m.sig <- pcemt.sig %>% subset(type=="Mesenchymal") %>% dplyr::pull(gene)

gene_sets4 <- list(PCEMT_EPITHELIAL = pcemt_e.sig,
                   PCEMT_MESENCHYMAL = pcemt_m.sig,
                   HIF_SIGNATURE = hif.sig
                   ) %>% append(mof_sigs) %>% append(bail_sigs)

## Oncogenic signatures (C6):
oncosigs = msigdbr(species = species, category = "C6")
gene_setsC6 <- oncosigs %>% split(x = .$gene_symbol, f = .$gs_name)

## Put all gene sets together for GSVA calculation:
gene_sets_all <- gene_sets1 %>% 
  append(gene_sets2) %>% # GO collections
  append(gene_sets3) %>% # TFT collections
  append(gene_sets4) %>% # pcEMT, HIF sig, fibroblast markers, etc.
  append(gene_setsC6)

hallhypoxia <- logtpm %>% filter(gene %in% gene_sets1$HALLMARK_HYPOXIA) %>% dplyr::select(-gene)
```



```{r Merging pcEMT and HIF signatures, warning=F, message=F}
## Merge HIF and pcEMT signature:
pcemt_hif_sig <- rbind(pcemt.sig, hif.sig_df)
rownames(pcemt_hif_sig) <- pcemt_hif_sig$gene

pcemt.hif <- pcemt_hif_sig$gene %>% match(logtpm$gene) %>% na.omit() %>% logtpm[.,]
rownames(pcemt.hif) <- pcemt.hif$gene

## Merge pcEMT and Hallmark Hypoxia signatures:
hallhyp_sig <- data.frame(gene=gene_sets1$HALLMARK_HYPOXIA, type="Hallmark Hypoxia")
pcemt_hallhyp_sig <- hallhyp_sig %>% 
  filter(gene %ni% pcemt.sig$gene) %>% 
  rbind(pcemt.sig)
 
```



# NMF
```{r pcEMT NMF calculations}
library(NMF)

## =============== Perform NMF calculations using pcEMT signature =============== ##
fn_nmfres <- "tcga_pcemt_nmf_results.Rdata" # file path for storing NMF results
perform_nmf <- F
nmf_method <- "brunet"
nmf_opts <- "vp8"
pcemt_4_nmf <- pcemt[,-1]
if(!file.exists(fn_nmfres) | perform_nmf){
  ## pcEMT NMF:
  nmf_pcemt <- nmf(pcemt_4_nmf, rank=2:9, method=nmf_method, nrun=50, .opt=nmf_opts, seed=123)
  nmf_pcemt2 <- nmf(pcemt_4_nmf, rank=2, method=nmf_method, nrun=500, .opt=nmf_opts, seed=123) # pcEMT, 2 clusters
  nmf_pcemt3 <- nmf(pcemt_4_nmf, rank=3, method=nmf_method, nrun=500, .opt=nmf_opts, seed=123) # pcEMT, 3 clusters
  
  ## Save NMF results for speed:
  nmfres_names <- c("nmf_pcemt","nmf_pcemt2","nmf_pcemt3")
  save(list = nmfres_names, file = fn_nmfres)
} else if(file.exists(fn_nmfres) & !exists("nmf_pcemt")){ 
  ## Load NMF results, if they haven't been loaded already:
  load(fn_nmfres) 
}
  
## Visualize NMF quality metrics:
plot(nmf_pcemt)
consensusmap(nmf_pcemt)
barplot(cophenetic*dispersion*silhouette.consensus~rank, data=nmf_pcemt$measures, main="NMF optimal factorization rank (pcEMT)")

## Visualize NMF results:
consensusmap(nmf_pcemt2)
consensusmap(nmf_pcemt3)

## =============== Get pcEMT NMF consensus clusters =============== ##

## Get data for heatmap:
pcemt_phm <- pcemt_4_nmf %>% t() %>% scale()
pcemt_df <- pcemt_phm %>% 
  data.frame(spl=rownames(.)) %>% 
  pivot_longer(-spl, names_to = "gene", values_to = "Expression z-score") %>% 
  inner_join(pcemt.sig, by="gene") %>% 
  dplyr::rename(pcEMT_sig = type)


## pcEMT, 2 NMF clusters:
nmf2_maxs <- coef(nmf_pcemt2) %>% apply(2, max)
nmf2_cc <- coef(nmf_pcemt2) %>% t()
for(i in 1:length(nmf2_maxs)){
  nmf2_cc[i,] <- nmf2_cc[i,] == nmf2_maxs[i]
}
nmf2_df <- nmf2_cc %>% 
  data.frame(spl=rownames(.), .) %>% 
  mutate(NMF2 = "",
         NMF2 = replace(NMF2, X1==1,"M-high"),
         NMF2 = replace(NMF2, X2==1,"M-low"),
         )


## pcEMT, 3 NMF clusters:
nmf3_maxs <- coef(nmf_pcemt3) %>% apply(2, max)
nmf3_cc <- coef(nmf_pcemt3) %>% t()
for(i in 1:length(nmf3_maxs)){
  nmf3_cc[i,] <- nmf3_cc[i,] == nmf3_maxs[i]
}
nmf3_df <- nmf3_cc %>% 
  data.frame(spl=rownames(.), .) %>% 
  mutate(NMF3 = "",
         NMF3 = replace(NMF3, X1==1,"E+/M+"),
         NMF3 = replace(NMF3, X2==1,"E-/M+"),
         NMF3 = replace(NMF3, X3==1,"E+/M-"),
         NMF3 = factor(NMF3, levels=c("E+/M+","E-/M+","E+/M-"))
         )


## Combine all NMF results:
nmf_cc_df <- nmf2_df %>%
  inner_join(nmf3_df, by="spl") %>%
  # inner_join(nmf2_df2, by="spl") %>%
  dplyr::select(contains(all_of(c("spl","NMF")))) %>% 
  inner_join(pcemt_df, by="spl")

## Add NMF cluster annotations to sample annotations:
spls <- spls %>%
  mutate(NMF2 = match(spl, nmf_cc_df$spl) %>% nmf_cc_df$NMF2[.],
         NMF2=factor(NMF2, levels=c("M-low","M-high")),
         NMF3 = match(spl, nmf_cc_df$spl) %>% nmf_cc_df$NMF3[.],
         NMF3 = factor(NMF3, levels=c("E+/M-","E-/M+","E+/M+")),
         )



## =============== Heatmaps w/ pcEMT signature and NMF clusters =============== ##
## Plot parameters:
len_out <- 6
breaks1 <- seq(-3,3,length.out=100)
breaks2 <- seq(-3,3,length.out=len_out+1)
colors1 <- paletteer_d("RColorBrewer::RdBu", 100, direction=-1, type="continuous")
annotcols <- paletteer_d("RColorBrewer::Set1")
annotcols2 <- paletteer_d("RColorBrewer::Set1")[c(4,5,3)]

## pcEMT, 2 clusters:
nmf_pcemt_hm2 <- nmf_cc_df %>%
  inner_join(spls, by="spl", suffix=c("",".x")) %>% 
  group_by(pcEMT_sig, NMF2) %>% 
  heatmap(.row=spl, .col=gene, .value=`Expression z-score`,
        show_column_dend=F, show_row_dend=F, show_row_names=F, .scale="none",
        cluster_columns=T, clustering_distance_columns="euclidean", clustering_method_columns="ward.D2",
        cluster_rows=T, clustering_distance_rows="euclidean", clustering_method_rows="ward.D2",
        palette_value=circlize::colorRamp2(breaks1, colors1),
        palette_grouping=list(annotcols2[c(1,3)], annotcols[2:1]),
        heatmap_legend_param=list(at = breaks2),
        column_title="TCGA PAAD: Pan-cancer EMT (pcEMT) signature mRNA expression (PDAC tumors only)", 
          column_title_gp=gpar(fontsize=8),
          row_title=NULL,
          column_names_gp=gpar(fontsize=8),
        use_raster=T,
        ) 
nmf_pcemt_hm2

# -- Save plot to PNG and PDF formats:
h <- 4
w <- 10.5
nmf_pcemt_hm2 %>% save_pdf("Images and plots/NMF2_pcEMT_sig_coef-clusters_clustergram.pdf", height=h, width=w)
png("Images and plots/NMF2_pcEMT_sig_coef-clusters_clustergram.png", height=h, width=w, units="in", res=500); nmf_pcemt_hm2; dev.off()

```


# GSVA and PROGENy
Here we perform gene set variation analysis (GSVA) to obtain enrichment scores for gene sets of interest.
```{r GSVA calculations}
library(GSVA)

#### Expression data prep ####
### Format data: Samples as columns, genes as rows
data_for_gsva <- logtpm[,-1] %>% data.matrix()
rownames(data_for_gsva) <- logtpm$gene

  
#### Calculate GSVA enrichment scores ####
force_calculate_gsva <- F
fn.gsva <- "gsva_results_PDAC-only.Rdata"
fn.gsva_csv <- "TCGA_gsva_results_PDAC-only.csv"
if(!file.exists(fn.gsva) | force_calculate_gsva){
  gsva_res <- gsva(
    data_for_gsva, 
    # gene_sets_all, 
    gene_sets_all %>% subset(str_detect(names(.),paste("HALLMARK_|KEGG_",paste(names(gene_sets4),collapse="|"),sep="|"))), 
    method = "gsva", 
    kcdf = "Gaussian",
    min.sz = 5, # Minimum number of genes required to include a gene set
    parallel.sz=detectCores()#-1
    )
  
  ### Save GSVA results:
  save(list = c("gsva_res"), file = fn.gsva)
  gsva_res %>%
    t() %>%
    data.frame(sample=rownames(.), .) %>% 
    write.csv(file=fn.gsva_csv, row.names=F) # save in CSV format for loading in Python
  
} else if (file.exists(fn.gsva) & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded already:
  load(fn.gsva) 
}

## Build data frame versions of GSVA results:
gsva_res_t.df <- gsva_res %>% t() %>% data.frame(spl=rownames(.))
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
rownames(gsva_res.df) <- gsva_res.df$gene.set
```

```{r PROGENy calculations}
library(progeny)
library(gglm)

## Calculate Progeny scores:
prog_res <- progeny(data_for_gsva, scale=T, top=100)
prog_res_df <- prog_res %>% data.frame(spl=rownames(.), .)

## Heatmap of Progeny scores:
Heatmap(prog_res)

## Hypoxia distribution:
ggplot(prog_res_df, aes(x=Hypoxia)) +
  geom_histogram(aes(y=..density..)) + 
  geom_vline(xintercept=median(prog_res_df$Hypoxia)) +
  geom_density()

## Linear model w/ Progeny scores:
prog_lm <- lm(Hypoxia ~ .+1, data=data.frame(prog_res)) %>% step(trace=0, direction="both")
summary(prog_lm)
gglm(prog_lm)
ggcoefstats(prog_lm, sort=T, stats.labels=F)
```


```{r Update sample annotations with GSVA and Progeny scores}
genesets_to_add <- c("PCEMT_EPITHELIAL",
                     "PCEMT_MESENCHYMAL",
                     "HIF_SIGNATURE",
                     names(gene_sets4),
                     "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION",
                     "HALLMARK_HYPOXIA"
                     ) %>% unique()
spls2 <- data.frame(spls, t(gsva_res)[,genesets_to_add]) %>% 
  inner_join(prog_res_df, by="spl")

```


## Compare Hypoxia enrichmenta across EMT clusters
```{r pcEMT comparisons}
line_size <- 0.6
bplotw <- 0.15
viow <- 0.75
w <- 3/1.5
h <- 4/1.5
pcemt_comp_cols <- annotcols2[c(3,1)]

## =============== Hallmark Hypoxia =============== ##
spls2 %>% 
ggbetweenstats(x = NMF2, y = HALLMARK_HYPOXIA,
               plot.type = "box", p.adjust.method="BH", type="np",
               point.args = list(alpha=0.7, shape=16, size=2,
                                 position = position_quasirandom(width = 0.15, groupOnX=T)),
               results.subtitle=T, centrality.plotting = F,
               ggplot.component = geom_boxplot(size=line_size, alpha=0, width=0.3),
               ggtheme = theme_cowplot(8, rel_small=1, rel_large=1, line_size=line_size),
               ) +
  scale_color_manual(values=pcemt_comp_cols) +
  labs(x="NMF cluster", y="Hallmark Hypoxia GSVA score", title="TCGA PAAD RNA-seq") + guides(color="none")
ggsave("Images and plots/HallmarkHypoxia-GSVA-score-tests_pcEMT-subtypes_PDAC-only.png", w=w*1.1, h=h)
ggsave("Images and plots/HallmarkHypoxia-GSVA-score-tests_pcEMT-subtypes_PDAC-only.pdf", w=w*1.1, h=h)

# -- Non-ggstatsplot version:
spls2 %>% 
  ggplot(aes(x = NMF2, y=HALLMARK_HYPOXIA, color=NMF2, fill=NMF2)) +
    geom_violin(color="black", size=line_size, width=viow) +
    geom_boxplot(color="black", fill="white", size=line_size, width=bplotw) +
    scale_color_manual(values=pcemt_comp_cols) +
    scale_fill_manual(values=pcemt_comp_cols) +
    theme_cowplot(8, rel_small=1, rel_large=1, line_size = line_size) + 
    labs(x="NMF cluster", y="Hallmark Hypoxia enrichment score", title="TCGA PAAD RNA-seq") + 
    guides(color="none", fill="none")
ggsave("Images and plots/HallmarkHypoxia-GSVA-score_violin_pcEMT-subtypes_PDAC-only.png", w=w, h=h)
ggsave("Images and plots/HallmarkHypoxia-GSVA-score_violin_pcEMT-subtypes_PDAC-only.pdf", w=w, h=h)



## =============== HIF signature =============== ##
spls2 %>% 
  ggbetweenstats(x = NMF2, y = HIF_SIGNATURE,
                 plot.type = "box", p.adjust.method="BH", type ="np", 
                 point.args = list(alpha=0.7, shape=16, size=2,
                                   position = position_quasirandom(width = 0.15, groupOnX=T)),
                 results.subtitle=T, centrality.plotting = F, 
                 ggplot.component = geom_boxplot(size=line_size, alpha=0, width=0.3),
                 ggtheme = theme_cowplot(8, rel_small=1, rel_large=1, line_size=line_size),
                 ) +
    scale_color_manual(values=pcemt_comp_cols) +
    labs(x="NMF cluster", y="HIF signature enrichment score", title="TCGA PAAD RNA-seq") + guides(color="none")
ggsave("Images and plots/HIF-GSVA-score-tests_pcEMT-subtypes_PDAC-only.png", w=w*1.1, h=h)
ggsave("Images and plots/HIF-GSVA-score-tests_pcEMT-subtypes_PDAC-only.pdf", w=w*1.1, h=h)

# -- Non-ggstatsplots version:
spls2 %>% 
  ggplot(aes(x = NMF2, y=HIF_SIGNATURE, color=NMF2, fill=NMF2)) +
    geom_violin(color="black", size=line_size, width=viow) +
    geom_boxplot(color="black", fill="white", size=line_size, width=bplotw) +
    scale_color_manual(values=pcemt_comp_cols) +
    scale_fill_manual(values=pcemt_comp_cols) +
    theme_cowplot(8, rel_small=1, rel_large=1, line_size = line_size) + 
    labs(x="NMF cluster", y="HIF signature GSVA score", title="TCGA PAAD RNA-seq") + 
    guides(color="none", fill="none")
ggsave("Images and plots/HIF-GSVA-score_violin_pcEMT-subtypes_PDAC-only.png", w=w, h=h)
ggsave("Images and plots/HIF-GSVA-score_violin_pcEMT-subtypes_PDAC-only.pdf", w=w, h=h)



## =============== Progeny Hypoxia pathway =============== ##
spls2 %>% 
  ggbetweenstats(x = NMF2, y = Hypoxia,
                 plot.type = "box", p.adjust.method="BH", type ="np", 
                 point.args = list(alpha=0.7, shape=16, size=2,
                                   position = position_quasirandom(width = 0.15, groupOnX=T)),
                 results.subtitle=T, centrality.plotting=F, 
                 ggplot.component = geom_boxplot(size=line_size, alpha=0, width=0.3),
                 ggtheme = theme_cowplot(8, rel_small=1, rel_large=1, line_size=line_size),
                 ) +
    scale_color_manual(values=pcemt_comp_cols) +
    labs(x="NMF cluster", y="Hypoxia PROGENy score", title="TCGA PAAD RNA-seq") + guides(color="none")
ggsave("Images and plots/Hypoxia-Progeny-score-tests_pcEMT-subtypes_PDAC-only.png", w=w*1.1, h=h)
ggsave("Images and plots/Hypoxia-Progeny-score-tests_pcEMT-subtypes_PDAC-only.pdf", w=w*1.1, h=h)

# -- Non-ggstatsplots version:
spls2 %>% 
  ggplot(aes(x = NMF2, y=Hypoxia, color=NMF2, fill=NMF2)) +
    geom_violin(color="black", size=line_size, width=viow) +
    geom_boxplot(color="black", fill="white", size=line_size, width=bplotw) +
    scale_color_manual(values=pcemt_comp_cols) +
    scale_fill_manual(values=pcemt_comp_cols) +
    theme_cowplot(8, rel_small=1, rel_large=1, line_size = line_size) + 
    labs(x="NMF cluster", y="Hypoxia PROGENy score", title="TCGA PAAD RNA-seq") + 
    guides(color="none", fill="none")
ggsave("Images and plots/Hypoxia-Progeny-score_violin_pcEMT-subtypes_PDAC-only.png", w=w, h=h)
ggsave("Images and plots/Hypoxia-Progeny-score_violin_pcEMT-subtypes_PDAC-only.pdf", w=w, h=h)
```


## PRCC calculations with hypoxia and immune/stromal content
```{r Compare immune and stromal signatures to pcEMT and HIF/Hypoxia enrichment}
w = 6/1.5
h = 4/1.5
lnsz <- 0.6
bw = 0.8
errh <- 0.25
pcorr_alpha <- 0.05 # significance level for (corrected) partial correlation p-values
bar_fill <- "grey50"

## =============== Partial rank correlation coefficient calculations =============== ##
## =============== Hallmark Hypoxia =============== ##
spls2 %>% 
  dplyr::select(contains(all_of(c("PCEMT_M","HALLMARK_HYPOX",".cell","fibroblast")))) %>%
  ggcorrmat(type="np", sig.level = pcorr_alpha, partial=T, output="dataframe") %>%
  filter(parameter1 %>% str_detect("PCEMT_M") | parameter2 %>% str_detect("PCEMT_M")) %>%
  transform(parameter2 = ifelse(parameter2 %>% str_detect("PCEMT_M"), parameter1, parameter2), # swap names if in wrong column
            parameter1 = ifelse(parameter2 %>% str_detect("PCEMT_M"), parameter2, parameter1)) %>% 
  mutate(
    parameter2 = gsub("\\."," ",parameter2),
    parameter2 = str_to_sentence(parameter2),
    parameter2 = case_when(str_detect(parameter2, "Hallmark") ~ "Hallmark Hypoxia",
                           str_detect(parameter2, "cd8") ~ "CD8+ T cell",
                           str_detect(parameter2, "cd4") ~ "CD4+ T cell",
                           str_detect(parameter2, "Nk") ~ "NK cell",
                           str_detect(parameter2, "Cancer") ~ "Cancer-associated fibroblast",
                           T ~ parameter2
                           )
  ) %>% 
  arrange(desc(estimate)) %>% mutate(parameter2=factor(parameter2, levels=rev(parameter2))) %>% 
# ggplot(aes(y=parameter2, x=estimate, fill = estimate > 0)) +
ggplot(aes(y=parameter2, x=estimate)) +
    geom_col(color="black", 
             fill=bar_fill,
             size=lnsz, width=bw) + 
    geom_vline(xintercept=0, size = lnsz) +
    geom_errorbarh(aes(xmin=conf.low, xmax=conf.high), size = lnsz, height = errh) +
    theme_cowplot(8, rel_small=1, rel_large=1, line_size=lnsz) + theme_cust(lnsz=lnsz*1.5) + 
    guides(col="none") +
    labs(
      title="Partial correlations with\npcEMT-M enrichment",
      x="Partial rank correlation coefficient", y=""
      )
ggsave("Images and plots/Partial-corrs_pcEMT-M_Hall-Hypoxia_PDAC-only.png", width=w, height=h)
ggsave("Images and plots/Partial-corrs_pcEMT-M_Hall-Hypoxia_PDAC-only.pdf", width=w, height=h)





## =============== HIF Signature =============== ##
spls2 %>% 
  dplyr::select(contains(all_of(c("PCEMT_M","HIF_SIG",".cell","fibroblast")))) %>%
  ggcorrmat(type="np", sig.level = pcorr_alpha, partial=T, output="dataframe") %>% 
  filter(parameter1 %>% str_detect("PCEMT_M") | parameter2 %>% str_detect("PCEMT_M")) %>%
  transform(parameter2 = ifelse(parameter2 %>% str_detect("PCEMT_M"), parameter1, parameter2), # swap names if in wrong column
            parameter1 = ifelse(parameter2 %>% str_detect("PCEMT_M"), parameter2, parameter1)) %>% 
  mutate(
    parameter2 = gsub("\\."," ",parameter2),
    parameter2 = str_to_sentence(parameter2),
    parameter2 = case_when(str_detect(parameter2, "Hif") ~ "HIF signature",
                           str_detect(parameter2, "cd8") ~ "CD8+ T cell",
                           str_detect(parameter2, "cd4") ~ "CD4+ T cell",
                           str_detect(parameter2, "Nk") ~ "NK cell",
                           str_detect(parameter2, "Cancer") ~ "Cancer-associated fibroblast",
                           T ~ parameter2
                           )
  ) %>% 
  arrange(desc(estimate)) %>% mutate(parameter2=factor(parameter2, levels=rev(parameter2))) %>% 
ggplot(aes(y=parameter2, x=estimate)) +
    geom_col(color="black", 
             fill=bar_fill,
             size=lnsz, width=bw) + 
    geom_vline(xintercept=0, size = lnsz) +
    geom_errorbarh(aes(xmin=conf.low, xmax=conf.high), size = lnsz, height = errh) +
    theme_cowplot(8, rel_small=1, rel_large=1, line_size=lnsz) + theme_cust(lnsz=lnsz*1.5) + 
    guides(col="none") +
    labs(
      title="Partial correlations with\npcEMT-M enrichment",
      x="Partial rank correlation coefficient", y=""
      )
ggsave("Images and plots/Partial-corrs_pcEMT-M_HIF_PDAC-only.png", width=w, height=h)
ggsave("Images and plots/Partial-corrs_pcEMT-M_HIF_PDAC-only.pdf", width=w, height=h)
```


# Survival analysis
Here we evaluate the survival probabilities of patients based on grouping tumors based on molecular subtypes/clusters.
```{r Load survival information}
## Load packages:
library(survival)
library(survminer)

## Add survival data to annotations for the tumors with RNA-seq data:
surv_data.inds <- match(spls$sample.id, surv_data$sample) 
spls_surv <- spls %>% 
  mutate(
    OS = surv_data$OS[surv_data.inds],
    OS.time = surv_data$OS.time[surv_data.inds],
    DSS = surv_data$DSS[surv_data.inds],
    DSS.time = surv_data$DSS.time[surv_data.inds],
    PFI = surv_data$PFI[surv_data.inds],
    PFI.time = surv_data$PFI.time[surv_data.inds],
    DFI = surv_data$DFI[surv_data.inds],
    DFI.time = surv_data$DFI.time[surv_data.inds],
    pcEMT_surv=NMF2,
    ) %>% 
  inner_join(spls2, by="sample.id", suffix=c("",".x")) %>%
  mutate(
    hypoxia_prog = as.character(as.numeric(Hypoxia>median(Hypoxia))+1),
    hypoxia_prog = case_when(hypoxia_prog==2 ~ "Hypoxia-high", hypoxia_prog==1 ~ "Hypoxia-low", T ~ hypoxia_prog),
    
    hallhyp = as.character(as.numeric(HALLMARK_HYPOXIA>median(HALLMARK_HYPOXIA))+1),
    hallhyp = case_when(hallhyp==2 ~ "Hypoxia-high", hallhyp==1 ~ "Hypoxia-low", T ~ hallhyp),
    
    pcemt_gsva = as.character(as.numeric(PCEMT_MESENCHYMAL>median(PCEMT_MESENCHYMAL))+1),
    pcemt_gsva = case_when(pcemt_gsva==2 ~ "M-high", pcemt_gsva==1 ~ "M-low", T ~ pcemt_gsva),
    
    pcemt_hallhyp = as.character(as.numeric(HALLMARK_HYPOXIA>median(HALLMARK_HYPOXIA) &
                                              PCEMT_MESENCHYMAL>median(PCEMT_MESENCHYMAL))+1),
    pcemt_hallhyp = case_when(pcemt_hallhyp==2 ~ "EMT/hypoxia-high", pcemt_hallhyp==1 ~ "EMT/hypoxia-low", T ~ pcemt_hallhyp),
  )

```


```{r Survival on GSVA score splits}
surv_colors <- paletteer_d("RColorBrewer::Set1")[3:5]
hypcomp_cols <- c("#4A97AF","#A34E6D")
plot_cis <- F
surv_xbreaks <- 500
w = 3.5
h = 3

## Set up survival object:
km2 <- Surv(time = spls_surv$DSS.time, event = spls_surv$DSS)

## =============== Hypoxia Progeny score split =============== ##
## Fit Kaplan-Meier, stratifying by M-high/low status:
km2_pcemt <- survfit(km2 ~ hypoxia_prog, data = spls_surv, type = "kaplan-meier", conf.type = "log")
km2_pcemt
survdiff(km2~hypoxia_prog, data = spls_surv) # Chi-squared test
ggsurvplot(km2_pcemt, conf.in=plot_cis, pval=T, break.x.by=surv_xbreaks, palette=hypcomp_cols[2:1],
           font.title=c(8,"bold"),
           font.x=8,
           font.y=8,
           font.tickslab=8,
           font.caption=8,
           font.subtitle=8,
           font.legend=8,
           pval.size=3,
           ) + 
  labs(title="Disease-specific survival (DSS)", x="Days")
ggsave("Images and plots/KM_Hypoxia-Progeny_DSS_PDAC-only.png", width=w, height=h)
ggsave("Images and plots/KM_Hypoxia-Progeny_DSS_PDAC-only.pdf", width=w, height=h)


## =============== HALLMARK HYPOXIA GSVA score split =============== ##
## Fit Kaplan-Meier, stratifying by M-high/low status:
km2_pcemt <- survfit(km2 ~ hallhyp, data = spls_surv, type = "kaplan-meier", conf.type = "log")
km2_pcemt
survdiff(km2~hallhyp, data = spls_surv) # Chi-squared test
ggsurvplot(km2_pcemt, conf.in=plot_cis, pval=T, break.x.by=surv_xbreaks, palette=hypcomp_cols[2:1],
           font.title=c(8,"bold"),
           font.x=8,
           font.y=8,
           font.tickslab=8,
           font.caption=8,
           font.subtitle=8,
           font.legend=8,
           pval.size=3,
           ) + 
  labs(title="Disease-specific survival (DSS)", x="Days")
ggsave("Images and plots/KM_HallHyp-GSVA_DSS_PDAC-only.png", width=w, height=h)
ggsave("Images and plots/KM_HallHyp-GSVA_DSS_PDAC-only.pdf", width=w, height=h)


```

```{r Survival on NMF clusters}
surv_colors <- paletteer_d("RColorBrewer::Set1")[3:5]
w = 3.5
h = 3

## Set up survival object:
km2 <- Surv(time = spls_surv$DSS.time, event = spls_surv$DSS)

## =============== pcEMT, 2 NMF clusters =============== ##
## Fit Kaplan-Meier, stratifying by M-high/low status:
km2_pcemt <- survfit(km2 ~ NMF2, data = spls_surv, type = "kaplan-meier", conf.type = "log")
km2_pcemt
survdiff(km2~NMF2, data = spls_surv) # Chi-squared test
ggsurvplot(km2_pcemt, conf.in=plot_cis, pval=T, break.x.by=surv_xbreaks, palette=surv_colors[1:2],
           font.title=c(8,"bold"),
           font.x=8,
           font.y=8,
           font.tickslab=8,
           font.caption=8,
           font.subtitle=8,
           font.legend=8,
           pval.size=3,
           ) + 
  labs(title="Disease-specific survival (DSS)", x="Days")
ggsave("Images and plots/KM_NMF-2_DSS_PDAC-only.png", width=w, height=h)
ggsave("Images and plots/KM_NMF-2_DSS_PDAC-only.pdf", width=w, height=h)

```


```{r pcEMT comps on Hallmark Hypoxia splits, warning=F, message=F}
line_size <- 0.6
w <- 3/1.5
h <- 4/1.5
hypcomp_cols <- c("#4A97AF","#A34E6D")

# -- ggstatsplot:
spls_surv %>% 
  mutate(
    hallhyp = case_when(str_detect(hallhyp,"high") ~ "High",
                        str_detect(hallhyp,"low") ~ "Low",
                        T ~ hallhyp,
                        ),
    hallhyp = factor(hallhyp, levels=c("Low","High")),
    ) %>%
  ggbetweenstats(x=hallhyp, y=PCEMT_MESENCHYMAL,
                 plot.type = "box", p.adjust.method="BH", type="np",
                 point.args = list(alpha=0.7, shape=16, size=2,
                                   position = position_quasirandom(width = 0.15, groupOnX=T)),
                 results.subtitle=T, centrality.plotting = T,
                 ggplot.component = geom_boxplot(size=line_size, alpha=0, width=0.3),
                 ggtheme = theme_cowplot(8, rel_small=1, rel_large=1, line_size=line_size),
                 ) +
  scale_color_manual(values=hypcomp_cols) + 
  labs(x="Hallmark Hypoxia", y="pcEMT-M enrichment score")
ggsave("Images and plots/pcEMT-comp_HallmarkHypoxia-median-split.png", w=w*1.2, h=h)
ggsave("Images and plots/pcEMT-comp_HallmarkHypoxia-median-split.pdf", w=w*1.2, h=h)

# -- Regulat ggplot:
spls_surv %>% 
  mutate(
    hallhyp = case_when(str_detect(hallhyp,"high") ~ "High",
                        str_detect(hallhyp,"low") ~ "Low",
                        T ~ hallhyp,
                        ),
    hallhyp = factor(hallhyp, levels=c("Low","High")),
    ) %>%
  ggplot(aes(x = hallhyp, y=PCEMT_MESENCHYMAL, fill=hallhyp)) +
    geom_violin(color="black", size=line_size, width=viow) +
    geom_boxplot(color="black", fill="white", size=line_size, width=bplotw) +
    scale_color_manual(values=hypcomp_cols) +
    scale_fill_manual(values=hypcomp_cols) +
    theme_cowplot(8, rel_small=1, rel_large=1, line_size = line_size) + 
    labs(x="Hallmark Hypoxia", y="pcEMT-M enrichment score", title="TCGA PAAD RNA-seq") + 
    guides(color="none", fill="none")
ggsave("Images and plots/HallmarkHypoxia-GSVA-score_violin_pcEMT-subtypes_PDAC-only.png", w=w, h=h)
ggsave("Images and plots/HallmarkHypoxia-GSVA-score_violin_pcEMT-subtypes_PDAC-only.pdf", w=w, h=h)
```


# Session info
```{r Print session info}
sink("sessionInfo.txt")
sessionInfo()
sink()
```