---
title: "Analysis of RNA-seq data from PDX tumors from UVA (Bauer Lab)"
author: Paul J. Myers
output:
  html_document:
    fig_width: 9
    fig_height: 6
    toc: yes
  pdf_document: 
    toc: yes
    fig_width: 9
    fig_height: 6
Date of creation: "1/06/2021"
R version: "4.1.2"
---

# Housekeeping Code

We start by loading some of the desired libraries for this script. Additional packages will be loaded later for specific analyses to make their use clear.

```{r Load packages, message=F, warning=F}
### Load packages:
library(tidyverse)
library(magrittr)
library(BiocManager)
library(paletteer)
library(cowplot)
library(ggrepel)
library(ggbeeswarm)
library(ggstatsplot)
library(Hmisc)
library(parallel)
library(org.Hs.eg.db)
```

```{r Clear workspace}
## Clear R workspace:
rm(list=ls())
```

```{r Check working directory}
## Check that current working directory is correct and the desired one:
cwd <- getwd()
cwd
```

```{r Define "not in" operator}
## Define 'not in' (ni) operator
`%ni%` <- Negate(`%in%`)

## Custom mode function:
getmode <- function(v) {
   uniqv <- unique(v)
   uniqv[which.max(tabulate(match(v, uniqv)))]
}
```

The function below can be used to generate a ggplot with a blank grid and black box surrounding the plot area.

```{r Custom ggplot theme for plot background}
theme_cust <- function(ticks=T, axis="y", box=T, lnsz=1, 
                       grid=F, grid_major=F, grid_minor=F, 
                       gridlnsz=0.5, gridlnsz_major=0.5, gridlnsz_minor=gridlnsz_major/2,
                       grid_linetype="dashed", grid_color="grey80",
                       xtickangle=0, hjust=NULL, vjust=NULL, fill=NA,
                       legend.position="right", strip_color=NA, strip_background=NA){
  theme_out <- theme(legend.position=legend.position, 
                     strip.background=element_rect(colour=strip_color, fill=strip_background),
                     )
  
  if (ticks == FALSE & axis == "y" ){
    theme_out <- theme_out + theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = fill),
      )
  }
  else if (ticks == FALSE & axis != "y"){
    theme_out <- theme_out + theme(
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = fill),
      )
  }
  else {
     theme_out <- theme_out + theme(
       axis.line.y = element_blank(),
       axis.line.x = element_blank(),
       panel.background = element_rect(size = lnsz, color = "black", fill = fill),
       axis.text.x = element_text(angle=xtickangle, hjust = hjust, vjust=vjust),
       ) 
  }
  if (!box) {
    theme_out <- theme_out + theme(panel.background = element_blank(),
                                   axis.line.y = element_line(size = lnsz, color = "black"),
                                   axis.line.x = element_line(size = lnsz, color = "black")
                                   )
  }
  # Grid lines:
  if ((grid_major & grid_minor) | grid){
    theme_out <- theme_out + 
      theme(panel.grid = element_line(color=grid_color,
                                      size = gridlnsz,
                                      linetype = grid_linetype)
            )
  } else if (grid_major & !grid_minor){
    theme_out <- theme_out + 
      theme(panel.grid.major = element_line(color=grid_color,
                                      size = gridlnsz_major,
                                      linetype = grid_linetype)
            )
  } else if (!grid_major & grid_minor){
    theme_out <- theme_out + 
      theme(panel.grid.minor = element_line(color=grid_color,
                                      size = gridlnsz_minor,
                                      linetype = grid_linetype)
            )    
  }
  return(theme_out)
}

# Julia's Plots.jl default color order:
julia_cols <- c("#009AFA","#E26E47","#3FA54E","#c270D2","#AD8F18","#01ABAE","#ED5F92","#C68324","#01A98C","#8F961E","#01A9CD","#9B7EE8","#618CF7","#F16072","#DC65B7","#6D9E33") %>% rep.int(4)
```

# Load and clean data

The data were provided in FPKM expression units with the gene lengths used to calculate these values. We'd like to use TPM (transcripts per million) instead, so we're going to convert the data into TPM from the RPKM values and the given gene lengths.

```{r Load data}
## Load RNA-seq and clinical/phenotype data:
library(readxl)
fn.fpkm <- "fpkm_genename2020.xlsx" # file with data in FPKM units
fn.spls <- "sample_types.csv" # file with sample type annotations
fn.clin <- "clinical_phenotypes.csv" # file with clinical phenotype annotations/data
data.all <- read_xlsx(fn.fpkm, skip = 1) %>%
  data.frame() # load RNA-seq data
spls <- read_csv(fn.spls) %>%
  data.frame(row.names = .$spl) # load sample types/annotation data
spls.pdx <- spls %>% 
  subset(PDX_line %ni% "None")
spls.tumors <- spls %>% 
  filter(spl_type == "Tumor")
clin_data <- read_csv(fn.clin) # load clinical/phenotype data
clin_data <- clin_data %>%
  mutate(PDX_line = match(spl,spls.pdx$spl) %>% spls.pdx$PDX_line[.]) # add associated PDX cell line names to clinical data


## Stelow's updated differentiation and tumor grading from 6/21/21:
fn_stelow_F0 <- "Stelow_F0_differentiation-stroma_6-21-21.csv"
fn_stelow_PDX <- "Stelow_PDX-tumor_differentiation-stroma_6-21-21.csv"
stelow_F0 <- read_csv(fn_stelow_F0) %>% # data for F0 tumors
  na.omit() %>% 
  separate(F0_spl_id, c("spl_id","F0_spl_id2")," ")
stelow_PDX <- read_csv(fn_stelow_PDX) %>% # data for PDX tumors
  na.omit() %>% 
  separate(PDX_spl_id, c("spl_id","PDX_spl_id2")," ")

# Merge Stelow differentiation/stroma data with rest of spl/phenotype data:
spls_stelow <- spls %>% 
  inner_join(stelow_F0, by="spl_id") %>% 
  inner_join(stelow_PDX, by="spl_id")
```

```{r Gene names and IDs}
## Load Entrez gene IDs that match ENSEMBL IDs:
fn.entrez <- "entrezIDs.csv"

if(file.exists(fn.entrez)){
  entrezIDs <- read.csv(fn.entrez)
} 

## Use biomaRt to get ENTREZ gene IDs if the file doesn't exist:
if(!exists("entrezIDs")){
  library(biomaRt)
  mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))

  entrezIDs <- getBM(
    filters="ensembl_gene_id",
    attributes=c("ensembl_gene_id", "entrezgene_id"),
    values=data.all$geneID,
    mart=mart)
  entrezIDs$entrezgene_id[entrezIDs$ensembl_gene_id=="ENSG00000227953"] <- 149134 # ENTREZ ID for LOC149134 gene
  write.csv(entrezIDs,"entrezIDs.csv",row.names = F)
}


## Get effective gene lengths:
gene_lengths <- data.all %>% 
  dplyr::select(geneID, Gene_name, GeneLength)
```

Now we convert the data from FPKM units to TPM units. As described in [Mortazavi et al.](https://doi.org/10.1038/nmeth.1226), RPKM/FPKM are calculated as $$ FPKM_i = \frac{X_i}{\tilde{l}_i N}*10^9 $$ where $X_i$ is the number of counts mapped to a feature (gene) $i$, $\tilde{l}_i$ is the effective gene length, and $N$ is the total number of counts within a sample.

The formula for TPM, which was originally described in [Li et al.](https://doi.org/10.1093/bioinformatics/btp692) is given by $$ TPM_i = \frac{X_i}{\tilde{l}_i}\left(\frac{1}{\Sigma_j\frac{X_j}{\tilde{l}_j }}\right) * 10^6 $$ where the $\Sigma_j$ is the summation of feature rates over all features within a sample.

Using these expressions and the relationships described in [this article](https://arxiv.org/abs/1104.3889v2) by Lior Pachter, it can be shown that TPM can be calculated from RPKM/FPKM according to $$ TPM_i = \left(\frac{FPKM_i}{\Sigma_j FPKM_j}\right) 10^6$$

```{r Clean the data and convert to TPM units}
 ################ CHOOSE WHETHER TO USE ONLY PROTEIN CODING GENES OR NOT ################
use_pc_genes_only <- T # TRUE = protein-coding genes only, FALSE = all available genes
if(use_pc_genes_only){
  ## Pull out just the protein-coding genes:
  data <- data.all %>% 
    subset(Gene_type %in% "protein_coding") 
  rownames(data) <- data$geneID
  
  ## Get just the FPKM values:
  fpkm <- data %>% 
    dplyr::select(-c(geneID, Gene_name, Chrom, Strand, Start, End, GeneLength, Gene_type, Description))
  rownames(fpkm) <- data$geneID

  ## Calculate TPM directly from FPKM values for the protein-coding genes only:
  tpm <- t(t(fpkm)/colSums(fpkm)*1e6) %>% data.frame() # Just the TPM values
  logtpm <- log2(tpm+1) # log2-transformed with pseudo-count 1 
  rownames(tpm) <- data$geneID # set row names using ENSEMBL gene IDs
  rownames(logtpm) <- data$geneID
  
} else {
  data <- data.all
  rownames(data) <- data$geneID
  
  ## Get just the FPKM values:
  fpkm <- data %>% 
    dplyr::select(-c(geneID, Gene_name, Chrom, Strand, Start, End, GeneLength, Gene_type, Description))
  rownames(fpkm) <- fpkm$geneID
  
  ## Calculate TPM directly from FPKM values using ALL genes in the data set:
  tpm <- t(t(fpkm)/colSums(fpkm)*1e6) %>% data.frame() # Just the TPM values
  logtpm <- log2(tpm+1) # log2-transformed with pseudo-count 1 
  rownames(tpm) <- data$geneID # set row names using ENSEMBL gene IDs
  rownames(logtpm) <- data$geneID
}

### Filter duplicate genes (based on gene symbol) by taking the Ensembl IDs with the highest variance:
gene_var <- apply(tpm, 1, var) %>% # calculate gene-wise variances
  data.frame(var=., geneID = names(.)) %>% # transform to data frame
  mutate(
    geneSymbol = match(geneID, data$geneID) %>% data$Gene_name[.] # get gene symbols
    ) %>% 
  group_by(geneSymbol) %>% 
  summarise(max(var), geneID = geneID) # select which Ensembl ID from each gene symbol has the highest variance 
  

### Filter the expression data with the selected Ensembl IDs from variance filtering:
tpm_filt <- tpm %>% 
  filter(rownames(.) %in% gene_var$geneID)

logtpm_filt <- logtpm %>% 
  filter(rownames(.) %in% gene_var$geneID)


### Data with gene symbols instead of Ensembl gene IDs:
tpm_symbol <- tpm_filt
rownames(tpm_symbol) <- rownames(tpm_symbol) %>% match(data$geneID) %>% data$Gene_name[.] %>% make.names(unique=T)
logtpm_symbol <- logtpm_filt # for log-transformed TPM values
rownames(logtpm_symbol) <- rownames(logtpm_symbol) %>% match(data$geneID) %>% data$Gene_name[.] %>% make.names(unique=T)
```

The data are now ready to be analyzed.


# Gene signatures of interest

Here we define the 13-gene (G13) signature from [Newhook et al.](https://doi.org/10.1371/journal.pone.0105631) and pull out the expression data for it.

```{r G13 signature}
g13.sig <- c("PRKCSH","PLCG1","RP11-439E19.3","CCDC88C","CUL3","ELAVL1","ULBP3","RIMKLB","CDH5","CD200R1","MS4A3","MDM2","TGFA") 
# g13.sig <- c("PRKCSH","PLCG1","LOC149134","CCDC88C","CUL3","ELAVL1","ULBP3","RIMKLB","CDH5","CD200R1","MS4A3","MDM2","TGFA") 

g13 <- match(g13.sig, data$Gene_name) %>% 
  logtpm[.,] %>% 
  data.frame(row.names = g13.sig) %>% 
  na.omit()
```

Next we introduce the "HIF signature" -- a set of 44 HIF transcription factor target genes that were described/used by [Li et al. (2014)](https://doi.org/10.1038/nature13557).

```{r HIF signature}
hif.sig <- c("IGFBP3", "EDN2", "PFKFB4", "FLT1", "TFR2", "BNIP3L", "TGFA","BNIP3","PGK1","EGLN1","LDHA","EGLN3","CP","TGFB3","PFKFB3",
             "HK1","TFRC","EDN1","CDKN1A","CA9","ADM1","HMOX1","SERPINE1","LOX","NDRG1","CA12","PDK1","VEGFA","ERO1L","RORA","P4HA1","MXI1",
             "SLC2A1","STC2","MIF","DDIT4","ENO1","CXCR4","PLOD1","P4HA2","GAPDH","PGAM1","TMEM45A","PIM1")

## TPM:
tpm.hif <- match(hif.sig, data$Gene_name) %>% na.omit() %>% tpm[.,]

## log2(TPM+1):
logtpm.hif <- match(hif.sig, data$Gene_name) %>% na.omit() %>% logtpm[.,]
```

An additional, 8-gene hypoxia gene signature is described by [Khouzam et al. (2021)](https://doi.org/10.3389/fimmu.2021.680435) and is predictive of survival in pancreatic cancer, as well as an immunosuppressed microenvironment, according to the authors.

```{r Hypoxia-8 gene set, message=F, warning=F}
hyp8_sig <- c("LDHA","SLC2A1","ANGPTL4","VEGFA","LOX","P4HA1","BNIP3","DDIT4")

## TPM:
tpm.hyp8 <- match(hyp8_sig, data$Gene_name) %>% na.omit() %>% tpm[.,]

## log2(TPM+1):
logtpm.hyp8 <- match(hyp8_sig, data$Gene_name) %>% na.omit() %>% logtpm[.,]
```


Classifying TCGA samples based on previously defined gene signatures: classical/epithelial, quasi-mesenchymal, and exocrine-like from the *Collisson PDAssigner signature*. The gene signatures were originally defined by [Collisson et al.](https://doi.org/10.1038/nm.2344) and were used by [Porter et al.](https://doi.org/10.1073/pnas.1914915116) to classify the TCGA PAAD samples. Since the latter publication does not have information on which samples were classified into which categories, we will reclassify the samples based on this signature.

```{r PDAC signatures, message=F}
## =============== Load the signatures =============== ##
# -- Collisson:
fn.coll <- "pdassigner_genes.csv"
coll.sig <- read_csv(fn.coll) %>% 
  mutate(sig = paste0("Collisson ",type))
coll_anno <- coll.sig[,-1] %>% data.frame() 
rownames(coll_anno) <- coll.sig$gene

# -- Moffitt:
fn.mof <- "Moffitt_PDAC_signature.txt" # PDAC subtype signature
fn.mof_strom <- "Moffitt_stromal_signature.txt" # stromal signature
mof.sig <- read_tsv(fn.mof) %>% 
  mutate(sig = paste0("Moffitt ",type))
mof_strom.sig <- read_tsv(fn.mof_strom) # Moffitt stromal signature


# -- Bailey:
# fn.bail <- "Bailey_PDAC_signature.txt"
# bail.sig <- read_tsv(fn.bail) %>% 
#   mutate(sig = paste0("Bailey ",type))

fn.bail <- "BaileySubtypeSig.csv"
bail.sig <- read_csv(fn.bail) %>% 
  mutate(
    gene = Symbol,
    type = "Bailey",
    sig = "Bailey"
    ) %>% 
  dplyr::select(gene, type, sig)

# -- Combine them:
pdac.sigs <- rbind(coll.sig, mof.sig, bail.sig)
```

We will also look at the immune and stromal signatures that underpin the ESTIMATE algorithm and see how their enrichment compares to the enrichment of the pcEMT signature (as well as how much overlap there is between pcEMT and these signatures). We can also compare the ESTIMATE Stromal scores from GSVA -- which we will calculate below -- to the stromal content measurements provided by Dr. Stelow's analysis.

```{r ESTIMATE immune and stromal signatures, message=F}
fn.estimate_sigs <- "../ESTIMATE_immune_stromal_signatures.txt"
est.sig <- read_tsv(fn.estimate_sigs)

est_strom <- est.sig %>% filter(type=="Stromal") %>% pull(gene)
est_imm <- est.sig %>% filter(type=="Immune") %>% pull(gene)

est_sigs <- list(ESTIMATE_IMMUNE_SIGNATURE = est_strom %>% mapIds(org.Hs.eg.db, ., "ENTREZID", "SYMBOL"),
                  ESTIMATE_STROMAL_SIGNATURE = est_imm %>% mapIds(org.Hs.eg.db, ., "ENTREZID", "SYMBOL")) # list for GSVA calculations

## Gene expression from ESTIMATE signatures:
est <- data$Gene_name %>%
  match(est.sig$gene) %>% na.omit() %>% 
  logtpm[.,]
```


```{r pcEMT signature, warning=F, message=F}
### Load pan-cancer EMT signature:
pcemt.fn <- "Pan-cancer-EMT-signature_Mak-et-al-2016.txt"
pcemt.sig <- read_tsv(pcemt.fn, col_names = c("gene","type"))
pcemt.sig.m <- read_tsv(pcemt.fn, col_names = c("gene","type")) %>% subset(type %in% "M")

### Get pcEMT signature expression data:
pcemt <- pcemt.sig$gene %>% match(data$Gene_name) %>% logtpm[.,]
pcemt.m <- pcemt.sig.m$gene %>% match(data$Gene_name) %>% logtpm[.,]
rownames(pcemt) <- pcemt.sig$gene
```

# Analysis
## GSVA
Now we load the necessary packages and gene sets from/related to the Molecular Signatures Database (MSigDB) for analyzing pathway and gene set enrichment in downstream analyses.

```{r Load packages and MSigDB collections, message=F, warning=F}
## Load packages:
library(GSVA)
library(msigdbr)


### Define MSigDB gene set collection(s) to use --> retrieve with 'msigdbr' package:
species = "Homo sapiens"
 
## Retrieve Hallmark and canonical pathways collections in the database (C1-H and C2-CP):
hall = msigdbr(species = species, category = "H")
cp = msigdbr(species = species, category = "C2", subcategory = "CP")
cp.b = msigdbr(species = species, category = "C2", subcategory = "CP:BIOCARTA")
cp.r = msigdbr(species = species, category = "C2", subcategory = "CP:REACTOME")
cp.p = msigdbr(species = species, category = "C2", subcategory = "CP:PID")
cp.k = msigdbr(species = species, category = "C2", subcategory = "CP:KEGG")
cp.w = msigdbr(species = species, category = "C2", subcategory = "CP:WIKIPATHWAYS")
cp_all <- rbind(cp, cp.b, cp.r, cp.p, cp.k, cp.w)
gene_sets1 <- rbind(hall, cp, cp.b, cp.r, cp.p, cp.k, cp.w) %>% split(x = .$entrez_gene, f = .$gs_name)
hallemt_sig <- hall %>% filter(gs_name=="HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% pull(gene_symbol) # Hallmark EMT genes
hallhyp_sig <- hall %>% filter(gs_name=="HALLMARK_HYPOXIA") %>% pull(gene_symbol) # Hallmark Hypoxia genes

## Go collections (C5):
go.bp <- msigdbr(species = species, category = "C5", subcategory = "GO:BP")
go.cc <- msigdbr(species = species, category = "C5", subcategory = "GO:CC")
go.mf <- msigdbr(species = species, category = "C5", subcategory = "GO:MF")
gene_sets2 <- rbind(go.bp, go.cc, go.mf) %>% split(x = .$entrez_gene, f = .$gs_name)

## Transcription factor target collections (C3-TFT):
tft <- msigdbr(species = species, category = "C3", subcategory = "TFT:GTRD")
tft_leg <- msigdbr(species = species, category = "C3", subcategory = "TFT:TFT_Legacy")
gene_sets3 <- rbind(tft, tft_leg) %>% split(x = .$entrez_gene, f = .$gs_name)

## Pan-cancer EMT sub-signatures and HIF signature:
pcemt_e.sig <- pcemt.sig %>% subset(type=="E") %>% dplyr::pull(gene) %>% 
  mapIds(org.Hs.eg.db, ., "ENTREZID", "SYMBOL")
pcemt_m.sig <- pcemt.sig %>% subset(type=="M") %>% dplyr::pull(gene) %>% 
  mapIds(org.Hs.eg.db, ., "ENTREZID", "SYMBOL")


gene_sets4 <- list(PCEMT_EPITHELIAL = pcemt_e.sig,
                   PCEMT_MESENCHYMAL = pcemt_m.sig,
                   HIF_SIGNATURE = hif.sig %>% mapIds(org.Hs.eg.db, ., "ENTREZID", "SYMBOL")
                   ) %>%
  append(est_sigs)

## Oncogenic signatures (C6):
oncosigs = msigdbr(species = species, category = "C6")
gene_setsC6 <- oncosigs %>% split(x = .$entrez_gene, f = .$gs_name)

## Put all gene sets together for GSVA calculation:
gene_sets_all <- gene_sets1 %>% 
  append(gene_sets2) %>% # GO collections
  append(gene_sets3) %>% # TFT collections
  append(gene_sets4) %>% # pcEMT, HIF sig, fibroblast markers, etc.
  append(gene_setsC6)


```

And now we calculate GSVA scores from the RNA-seq data.

```{r GSVA calculation}
#### Expression data prep ####
### Format data: Samples as columns, genes as rows
data_for_gsva <- logtpm %>% data.matrix()
rownames(data_for_gsva) <- rownames(logtpm) %>% mapIds(org.Hs.eg.db, ., "ENTREZID", "ENSEMBL")

#### Calculate GSVA scores ####
fn_pdx <- "PDX_gsva_results.Rdata"
force_calculate_gsva <- F
if(!file.exists(fn_pdx) | force_calculate_gsva){
  gsva_res <- gsva(
    data_for_gsva, 
    gene_sets_all, 
    method = "gsva", 
    kcdf = "Gaussian",
    mx.diff = T,
    min.sz = 5, # Minimum number of genes required to include a gene set
    # abs.ranking = F,
    parallel.sz=detectCores()#-1
    )
  
  ## Save GSVA results:
  save(list = c("gsva_res"), file = fn_pdx)
  
} else if (file.exists(fn_pdx) & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded already:
  load(fn_pdx) 
}
## Build data frame versions of GSVA results:
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
rownames(gsva_res.df) <- gsva_res.df$gene.set
```


# Analysis of tumor differentiation and EMT status
We are interested in determining the correlation between the molecular characterization of a tumor's mesenchymal status/state and tumor differentiation status. We will also look at correlations between stromal content as measured by Dr. Stelow at UVA and the GSVA scores for the ESTIMATE Stromal and Immune signatures.
## All tumors

```{r Collect and format phenotype data}
## Tumor grading (histologic, pathologic, etc.) and pcEMT-M scores:
diff_data <- clin_data %>% 
  dplyr::select(spl,
                contains(all_of(c("differentiation"))),
                # contains(all_of(c("differentiation","stage"))),
                starts_with("path_"),
                -contains("Stelow"),
                ) %>% 
  mutate(
    pcEMT.M = spl %>% match(colnames(gsva_res)) %>% gsva_res["PCEMT_MESENCHYMAL",.],
    Hallmark_EMT = spl %>% match(colnames(gsva_res)) %>% gsva_res["HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION",.]
    ) 

## Longer version of the data:
diff_data.l <- diff_data %>% 
  pivot_longer(-c(spl, pcEMT.M, Hallmark_EMT), values_to = "grade", names_to = "grading_metric") %>% 
  mutate(
    grade = factor(grade, levels = c("NR","well","well to poor","well to moderate",
                                     "moderate","moderate to poor","poor",
                                     "1 to 3","2","2 to 3","3",
                                     "PRIMARY","MET"
    ))
)
```

And now we assess the relationships between pcEMT-M enrichment and tumor grading metrics.

```{r EMT-tumor grade comparisons, message=F, warning=F}
## Data wrangling:
stelow_emt <- spls_stelow %>% inner_join(emt_gsva.t, by="spl") %>% 
  group_by(spl_id) %>% 
  mutate(
    F0_pct_stroma = mean(F0_pct_stroma_Stelow),
    PDX_pct_stroma = mean(PDX_pct_stroma_Stelow),
    F0_differentiation = getmode(F0_differentiation_Stelow),
    PDX_differentiation = getmode(PDX_differentiation_Stelow),
    ) %>% 
  dplyr::select(-contains(all_of(c("_Stelow","spl_id2")))) %>% 
  unique() %>% 
  mutate(PDX_line = replace(PDX_line, PDX_line=="None", NA))

jitter_pos <- position_quasirandom(width = 0.15, groupOnX=T) # for setting positions of points
w = 2.5
h = 3
lblsz = 2; rpldir = "x"
rpla = 1; rplf = 50
fntsz = 8

## ============== F0 differentiation, pcEMT-M ============== ##
type <- "np"
if (type=="np") {file_end <- "MannWhitneyU_"} else {file_end <- ""}
stelow_emt %>% 
  ggbetweenstats(x=F0_differentiation, y=PCEMT_MESENCHYMAL,
                 plot.type = "box",
                 violin.args=list(size=1),
                 point.args = list(alpha=1, shape=16, size=2, position = jitter_pos),
                 centrality.plotting=F,
                 type = type,
                 ggtheme = theme_cowplot(fntsz, line_size=0.5, rel_small=1, rel_tiny=1, rel_large=1),
                 ggplot.component = list( # extra ggplot geoms
                   geom_boxplot(fill=NA, size=1, alpha=0.3, width=0.3, outlier.alpha=0),
                   # Label the tumors with associated PDX cell lines:
                   geom_label_repel(aes(label=.$PDX_line, fill=.$PDX_line), alpha=rpla, force=rplf,
                                    min.segment.length = 0, size=lblsz, position=jitter_pos, direction=rpldir
                                    )
                   )
                 ) +
    scale_color_brewer(palette="Set1", direction=-1) +
    scale_fill_brewer(palette="Dark2") +
    #theme(axis.text.x = element_text(angle=60, hjust=1)) +
    labs(x="F0 differentiation", y="pcEMT-M enrichment score", title="Bauer Lab PDX RNA-seq samples") +
    guides(color=F)
ggsave(paste("F0_differentiation-pcEMT_comp_",file_end,"Stelow.png",sep=""), width=w, height=h)
ggsave(paste("F0_differentiation-pcEMT_comp_",file_end,"Stelow.pdf",sep=""), width=w, height=h)

## ============== F0 differentiation, Hallmark EMT ============== ##
type <- "np"
if (type=="np") {file_end <- "MannWhitneyU_"} else {file_end <- ""}
stelow_emt %>% 
  ggbetweenstats(x=F0_differentiation, y=HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION,
                 plot.type = "box",
                 violin.args=list(size=1),
                 point.args = list(alpha=1, shape=16, size=3, position = jitter_pos),
                 centrality.plotting=F,
                 type = type,
                 ggtheme = theme_cowplot(fntsz, line_size=0.5, rel_small=1, rel_tiny=1, rel_large=1),
                 ggplot.component = list( # extra ggplot geoms
                   geom_boxplot(fill=NA, size=1, alpha=0.3, width=0.3, outlier.alpha=0),
                   # Label the tumors with associated PDX cell lines:
                   geom_label_repel(aes(label=.$PDX_line, fill=.$PDX_line), alpha=rpla, force=rplf,
                                    min.segment.length = 0, size=lblsz, position=jitter_pos, direction=rpldir
                                    )
                   )
                 ) +
    scale_color_brewer(palette="Set1", direction=-1) +
    scale_fill_brewer(palette="Dark2") +
    #theme(axis.text.x = element_text(angle=60, hjust=1)) +
    labs(x="F0 differentiation", y="Hallmark EMT enrichment score", title="Bauer Lab RNA-seq samples") + guides(color=F)
ggsave(paste("F0_differentiation-HallEMT_comp_",file_end,"Stelow.png",sep=""), width=w, height=h)
ggsave(paste("F0_differentiation-HallEMT_comp_",file_end,"Stelow.pdf",sep=""), width=w, height=h)


## ============== PDX tumor differentiation, pcEMT-M ============== ##
stelow_emt %>% 
  ggbetweenstats(x=PDX_differentiation, y=PCEMT_MESENCHYMAL,
                 plot.type = "box",
                 violin.args=list(size=1),
                 point.args = list(alpha=1, shape=16, size=3, position = jitter_pos),
                 centrality.plotting=F,
                 type = type,
                 ggtheme = theme_cowplot(fntsz, line_size=0.5, rel_small=1, rel_tiny=1, rel_large=1),
                 ggplot.component = list( # extra ggplot geoms
                   geom_boxplot(fill=NA, size=1, alpha=0.3, width=0.3, outlier.alpha=0),
                   # Label the tumors with associated PDX cell lines:
                   geom_label_repel(aes(label=.$PDX_line, fill=.$PDX_line), alpha=rpla, force=rplf,
                                    min.segment.length = 0, size=lblsz, position=jitter_pos, direction=rpldir
                                    )
                   )
                 ) +
    scale_color_brewer(palette="Set1", direction=-1) +
    scale_fill_brewer(palette="Dark2") +
    #theme(axis.text.x = element_text(angle=60, hjust=1)) +
    labs(x="PDX differentiation", y="pcEMT-M enrichment score", title="Bauer Lab RNA-seq samples") + guides(color=F)
ggsave(paste("PDX_differentiation-pcEMT_comp_",file_end,"Stelow.png",sep=""), width=w, height=h)
ggsave(paste("PDX_differentiation-pcEMT_comp_",file_end,"Stelow.pdf",sep=""), width=w, height=h)


## ============== PDX tumor differentiation, Hallmark EMT ============== ##
stelow_emt %>% 
  ggbetweenstats(x=PDX_differentiation, y=HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION,
                 plot.type = "box",
                 violin.args=list(size=1),
                 point.args = list(alpha=1, shape=16, size=3, position = jitter_pos),
                 centrality.plotting=F,
                 type = type,
                 ggtheme = theme_cowplot(fntsz, line_size=0.5, rel_small=1, rel_tiny=1, rel_large=1),
                 ggplot.component = list( # extra ggplot geoms
                   geom_boxplot(fill=NA, size=1, alpha=0.3, width=0.3, outlier.alpha=0),
                   # Label the tumors with associated PDX cell lines:
                   geom_label_repel(aes(label=.$PDX_line, fill=.$PDX_line), alpha=rpla, force=rplf,
                                    min.segment.length = 0, size=lblsz, position=jitter_pos, direction=rpldir
                                    )
                   )
                 ) +
    scale_color_brewer(palette="Set1", direction=-1) +
    scale_fill_brewer(palette="Dark2") +
    #theme(axis.text.x = element_text(angle=60, hjust=1)) +
    labs(x="PDX differentiation", y="Hallmark EMT enrichment score", title="Bauer Lab RNA-seq samples") +
    guides(color=F)
ggsave(paste("PDX_differentiation-HallEMT_comp_",file_end,"Stelow.png",sep=""), width=w, height=h)
ggsave(paste("PDX_differentiation-HallEMT_comp_",file_end,"Stelow.pdf",sep=""), width=w, height=h)
```


## Only tumors with unchanged F0-PDX/F1 grades
```{r Differentiation analysis, message=F, warning=F}
stelow_emt2 <- stelow_emt %>% filter(F0_differentiation==PDX_differentiation) # filter for F0_diff=PDX_diff
jitter_pos <- position_quasirandom(width = 0.15, groupOnX=T) # for setting positions of points
w = 3.1
h = 4
lblsz = 2.5; rpldir = "x"
rpla = 1; rplf = 50
fntsz = 12

## Welch's t-test:
stelow_emt2 %>% 
  # ggbetweenstats(x=F0_differentiation, y=PCEMT_MESENCHYMAL,
  ggbetweenstats(x=F0_differentiation, y=HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION,
                 plot.type = "box",
                 violin.args=list(size=1),
                 point.args = list(alpha=1, shape=16, size=3, position = jitter_pos),
                 centrality.plotting=F,
                 type = "p",
                 ggtheme = theme_cowplot(fntsz, line_size=0.8, rel_small=1, rel_tiny=1, rel_large=1),
                 ggplot.component = list( # extra ggplot geoms
                   geom_boxplot(fill=NA, size=1, alpha=0.3, width=0.3, outlier.alpha=0),
                   # Label the tumors with associated PDX cell lines:
                   geom_label_repel(aes(label=.$PDX_line, fill=.$PDX_line), alpha=rpla, force=rplf,
                                    min.segment.length = 0, size=lblsz, position=jitter_pos, direction=rpldir
                                    )
                   )
                 ) +
    scale_color_brewer(palette="Set1", direction=-1) +
    scale_fill_brewer(palette="Dark2") +
    #theme(axis.text.x = element_text(angle=60, hjust=1)) +
    # labs(x="F0 differentiation (Stelow)", y="pcEMT-M GSVA score", title="Bauer Lab RNA-seq samples: F0=F1/PDX") + guides(color=F)
    labs(x="F0 differentiation (Stelow)", y="Hallmark EMT GSVA score", title="Bauer Lab RNA-seq samples: F0=F1/PDX") + guides(color=F)
# ggsave("F0=F1_differentiation-pcEMT_comp_Stelow.png", width=w, height=h)
ggsave("F0=F1_differentiation-HallEMT_comp_Stelow.png", width=w, height=h)


### Mann-Whitney U test:
## =============== pcEMT-M ============== ##
stelow_emt2 %>% 
  ggbetweenstats(x=F0_differentiation, y=PCEMT_MESENCHYMAL,
                 plot.type = "box",
                 violin.args=list(size=1),
                 point.args = list(alpha=1, shape=16, size=3, position = jitter_pos),
                 centrality.plotting=F,
                 type = "np",
                 ggtheme = theme_cowplot(fntsz, line_size=0.8, rel_small=1, rel_tiny=1, rel_large=1),
                 ggplot.component = list( # extra ggplot geoms
                   geom_boxplot(fill=NA, size=1, alpha=0.3, width=0.3, outlier.alpha=0),
                   # Label the tumors with associated PDX cell lines:
                   geom_label_repel(aes(label=.$PDX_line, fill=.$PDX_line), alpha=rpla, force=rplf,
                                    min.segment.length = 0, size=lblsz, position=jitter_pos, direction=rpldir
                                    )
                   )
                 ) +
    scale_color_brewer(palette="Set1", direction=-1) +
    scale_fill_brewer(palette="Dark2") +
    #theme(axis.text.x = element_text(angle=60, hjust=1)) +
    labs(x="F0 differentiation (Stelow)", y="pcEMT-M GSVA score", title="Bauer Lab RNA-seq samples: F0=F1/PDX") + guides(color=F)
ggsave("F0=F1_differentiation-pcEMT_comp_MannWhitneyU_Stelow.png", width=w, height=h)


## =============== Hallmark EMT ============== ##
stelow_emt2 %>% 
  ggbetweenstats(x=F0_differentiation, y=HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION,
                 plot.type = "box",
                 violin.args=list(size=1),
                 point.args = list(alpha=1, shape=16, size=3, position = jitter_pos),
                 centrality.plotting=F,
                 type = "np",
                 ggtheme = theme_cowplot(fntsz, line_size=0.8, rel_small=1, rel_tiny=1, rel_large=1),
                 ggplot.component = list( # extra ggplot geoms
                   geom_boxplot(fill=NA, size=1, alpha=0.3, width=0.3, outlier.alpha=0),
                   # Label the tumors with associated PDX cell lines:
                   geom_label_repel(aes(label=.$PDX_line, fill=.$PDX_line), alpha=rpla, force=rplf,
                                    min.segment.length = 0, size=lblsz, position=jitter_pos, direction=rpldir
                                    )
                   )
                 ) +
    scale_color_brewer(palette="Set1", direction=-1) +
    scale_fill_brewer(palette="Dark2") +
    #theme(axis.text.x = element_text(angle=60, hjust=1)) +
    labs(x="F0 differentiation (Stelow)", y="Hallmark EMT GSVA score", title="Bauer Lab RNA-seq samples: F0=F1/PDX") + guides(color=F)
ggsave("F0=F1_differentiation-HallEMT_comp_MannWhitneyU_Stelow.png", width=w, height=h)
```


# Session info
```{r Print session info}
sink("sessionInfo_PDX-RNAseq.txt")
sessionInfo()
sink()
``` 
